.github/workflows/main.yml

```diff
@@ -1,79 +1,256 @@
-
-    - name: set up environment
-      run: ./ci/setup_env.sh "${{ matrix.python_version }}" "$BACKENDS"
-      shell: bash
+      - name: checkout
+        uses: actions/checkout@v2
 
-    - name: run tests
-      run: PYTEST_BACKENDS=$BACKENDS ./ci/run_tests.sh --numprocesses auto
-      shell: bash
+      - name: install python
+        uses: actions/setup-python@v2
 
-    - name: pytest errors
-      run: ./ci/pytest_errors.sh
-      if: failure()
+      - name: install poetry
+        run: pip install poetry-core black tomli
 
-  Tests_dask:
-    name: Tests dask
+      - name: check whether setup.py needs to be regenerated
+        env:
+          # this environment variable is set is to ensure reproducible setup.py
+          # generation
+          #
+          # because various data structures in poetry are unordered, an can
+          # result in things like package extras appearing in a different order
+          PYTHONHASHSEED: 0
+        run: |
+          set -euo pipefail
+
+          python ./dev/poetry2setup.py -o setup.py
+          git diff --exit-code setup.py
+
+  gen_conda_env_file:
     runs-on: ubuntu-latest
-    env:
-      BACKENDS: "dask"
-    strategy:
-      fail-fast: false
-      matrix:
-        python_version: ["3.7", "3.9"]
+    needs:
+      - check_setup_py
     steps:
-    - name: checkout
-      uses: actions/checkout@v1
+      - name: checkout
+        uses: actions/checkout@v2
+
+      - name: install python
+        uses: actions/setup-python@v2
+
+      - name: install poetry2conda
+        run: pip install poetry2conda
 
-    - name: set up environment
-      run: ./ci/setup_env.sh "${{ matrix.python_version }}" "$BACKENDS"
-      shell: bash
+      - name: generate conda env file
+        run: |
+          set -euo pipefail
 
-    - name: run tests
-      run: PYTEST_BACKENDS=$BACKENDS ./ci/run_tests.sh --numprocesses auto
-      shell: bash
+          poetry2conda --dev --extras all pyproject.toml - | tee environment.yaml
 
-    - name: pytest errors
-      run: ./ci/pytest_errors.sh
-      if: failure()
+      - name: upload conda env file
+        uses: actions/upload-artifact@v2
+        with:
+          name: conda-env-file
+          path: environment.yaml
+          if-no-files-found: error
 
-  Tests_sql:
-    name: Tests SQL
+      - name: cleanup generated env files
+        run: rm environment.yaml
+
+  test_no_backends:
+    runs-on: ${{ matrix.os }}
+    needs:
+      - gen_conda_env_file
+    defaults:
+      run:
+        shell: bash -l {0}
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-latest
+          - windows-latest
+        python-version:
+          - "3.7"
+          - "3.9"
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+
+      - name: download env file
+        uses: actions/download-artifact@v2
+        with:
+          name: conda-env-file
+          path: tmp
+
+      - uses: conda-incubator/setup-miniconda@v2
```

Comment:
Since we have a concise poetry setup now, I'd just remove conda from the CI setup altogether. We can install all of the dependencies here from PyPI.

Reply:
I'd also remove the `gen_conda_env_file` job making the setup much simpler.

Reply:
Setting up miniconda and installing pyarrow in a separate step [takes 7 minutes](https://github.com/ibis-project/ibis/runs/3535730775) even with mamba.

Reply:
Poetry install without a lockfile took 420s on my local machine while poetry install with a lock file took less than 15 seconds to install.

Reply:
Experimented further with poetry. Seems like `poetry update` is really quick after having a lockfile.

My proposal is either:
- to have a poetry lockfile in the repository, keep it updated using dependabot and always call poetry update before install: `poetry update && poetry install` OR
- install ibis using pip: `pip install -e ibis` 

This way the CI jobs will use the latest dependencies.

Reply:
Yep, I've almost got this working and installation is much faster

Reply:
Done.